File: backend/app/schemas/book.py
from __future__ import annotations

from typing import List, Optional

from pydantic import BaseModel, Field


class AuthorBrief(BaseModel):
    id: int
    name: str
    role: str | None = None

    model_config = {"from_attributes": True}


class FormatBrief(BaseModel):
    id: int
    format: str
    file_size: int | None = None
    mime_type: str | None = None

    model_config = {"from_attributes": True}


class IdentifierBrief(BaseModel):
    id: int
    type: str
    value: str

    model_config = {"from_attributes": True}


class TagBrief(BaseModel):
    id: int
    name: str
    color: str | None = None

    model_config = {"from_attributes": True}


class CategoryBrief(BaseModel):
    id: int
    name: str

    model_config = {"from_attributes": True}


# ---------------------------------------------------------------------------
# List item (lightweight, used in paginated listing)
# ---------------------------------------------------------------------------


class BookListItem(BaseModel):
    id: int
    title: str
    subtitle: str | None = None
    cover_path: str | None = None
    language: str | None = None
    page_count: int | None = None
    published_date: str | None = None
    rating: float | None = None
    progress: float | None = None
    authors: List[AuthorBrief] = []
    formats: List[FormatBrief] = []
    tags: List[TagBrief] = []
    created_at: str
    updated_at: str

    model_config = {"from_attributes": True}


# ---------------------------------------------------------------------------
# Detail (full book info)
# ---------------------------------------------------------------------------


class BookDetail(BaseModel):
    id: int
    title: str
    sort_title: str | None = None
    subtitle: str | None = None
    description: str | None = None
    cover_path: str | None = None
    language: str | None = None
    page_count: int | None = None
    published_date: str | None = None
    rating: float | None = None
    rating_source: str | None = None
    metadata_source: str | None = None
    series_index: float | None = None
    created_at: str
    updated_at: str

    authors: List[AuthorBrief] = []
    formats: List[FormatBrief] = []
    identifiers: List[IdentifierBrief] = []
    tags: List[TagBrief] = []
    categories: List[CategoryBrief] = []

    publisher: dict | None = None
    series: dict | None = None

    model_config = {"from_attributes": True}


# ---------------------------------------------------------------------------
# Pagination
# ---------------------------------------------------------------------------


class PaginationInfo(BaseModel):
    page: int = 1
    page_size: int = 20
    total_items: int = 0
    total_pages: int = 0


class BookListResponse(BaseModel):
    items: List[BookListItem] = []
    pagination: PaginationInfo

File: frontend/src/types/book.ts
export interface AuthorBrief {
  id: number
  name: string
  role?: string
}

export interface FormatBrief {
  id: number
  format: string
  file_size: number | null
  mime_type: string | null
}

export interface TagBrief {
  id: number
  name: string
  color?: string
}

export interface Book {
  id: number
  title: string
  subtitle?: string
  sort_title?: string
  description?: string
  cover_path?: string
  language?: string
  page_count?: number
  published_date?: string
  rating?: number
  progress?: number | null
  rating_source?: string
  metadata_source?: string
  series_index?: number
  // These come from backend as arrays of objects
  authors: AuthorBrief[]
  formats: FormatBrief[]
  tags: TagBrief[]
  // Optional nested
  publisher?: { id: number; name: string }
  series?: { id: number; name: string }
  categories?: Array<{ id: number; name: string }>
  identifiers?: Array<{ id: number; type: string; value: string }>
  // Legacy compat for mock data
  isbn?: string
  has_cover?: boolean
  created_at: string
  updated_at: string
}

export interface PaginationInfo {
  total_items: number
  total_pages: number
  page: number
  page_size: number
}

export interface BookListResponse {
  items: Book[]
  pagination: PaginationInfo
}

export interface Stats {
  total_books: number
  total_authors: number
  total_tags: number
  total_series: number
  total_categories: number
  total_publishers: number
  total_shelves: number
}
